boa/src/builtins/object/mod.rs

```diff
@@ -877,3 +913,48 @@ fn object_define_properties(
     // 7. Return O.
     Ok(())
 }
+
+/// Type enum used in the abstract operation GetOwnPropertyKeys
+#[derive(Debug, Copy, Clone)]
+enum GetOwnPropertyKeysType {
+    String,
+    Symbol,
+}
+
+/// The abstract operation GetOwnPropertyKeys
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-getownpropertykeys
+fn get_own_property_keys(
+    o: &JsValue,
+    r#type: GetOwnPropertyKeysType,
+    context: &mut Context,
+) -> JsResult<JsValue> {
+    // 1. Let obj be ? ToObject(o).
+    let obj = o.to_object(context)?;
+
+    // 2. Let keys be ? obj.[[OwnPropertyKeys]]().
+    let keys = obj.__own_property_keys__(context)?;
+
+    // 3. Let nameList be a new empty List.
+    // 4. For each element nextKey of keys, do
+    let name_list: Vec<_> = keys
+        .iter()
```

Comment:
Would it be possible to use `into_iter` to avoid the `clone`s inside the `filter_map`?

Reply:
Yeah that makes sense!

Reply:
I just tried to make the change, but I'm not sure if this is possible. I tried changing to the following:

```rust
let name_list = keys.into_iter().filter_map(|next_key| {
    // a. If Type(nextKey) is Symbol and type is symbol or Type(nextKey) is String and type is string, then
    // i. Append nextKey as the last element of nameList.
    match (r#type, next_key) {
        (PropertyKeyType::String, PropertyKey::String(key)) => Some(key.into()),
        (PropertyKeyType::String, PropertyKey::Index(index)) => Some(index.into()),
        (PropertyKeyType::Symbol, PropertyKey::Symbol(symbol)) => Some(symbol.into()),
        _ => None,
    }
});
```

But then I get the error: ```[rustc E0509] [E] cannot move out of type `property::PropertyKey`, which implements the `Drop` trait```.

Am I missing something here? I'm rather new to rust 😅

Reply:
That's because we need to run a custom `Drop` implementation (the destructor) for `PropertyKey` in order to be Gc'd, meaning we cannot destructure it, as that would make the `Drop` call reference null memory. This is a bug we have right now, because a PropertyKey shouldn't be garbage collected.
However, you can change the match to:
```Rust
match (r#type, &next_key) {
        (PropertyKeyType::String, PropertyKey::String(_)) => Some(next_key.into()),
        (PropertyKeyType::String, PropertyKey::Index(_)) => Some(next_key.into()),
        (PropertyKeyType::Symbol, PropertyKey::Symbol(_)) => Some(next_key.into()),
        _ => None,
    }
```
and it should work, I think.

Reply:
Ah I see, gotcha. Your suggestion doesn't work I think, because when `next_key` matches `PropertyKey::Index`, we want to explicitly convert the integer index to string, instead of constructing `JsValue::Integer`.

Reply:
There's a similar implementation in https://github.com/boa-dev/boa/blob/aacdd12ed900a1a0db6debd478f303422885a831/boa/src/object/operations.rs#L516-L520, and it looks like a copy/clone is made 🤔.

Reply:
On that example a `clone` is necessary because `key` is reused after the conversion

Reply:
> Ah I see, gotcha. Your suggestion doesn't work I think, because when `next_key` matches `PropertyKey::Index`, we want to explicitly convert the integer index to string, instead of constructing `JsValue::Integer`.

Ah, I forgot that 😅
Then this should work:

```Rust
match (r#type, &next_key) {
        (PropertyKeyType::String, PropertyKey::String(_)) => Some(next_key.into()),
        (PropertyKeyType::String, PropertyKey::Index(index)) => Some(*index.to_string().into()),
        (PropertyKeyType::Symbol, PropertyKey::Symbol(_)) => Some(next_key.into()),
        _ => None,
    }
```

Reply:
> On that example a `clone` is necessary because `key` is reused after the conversion

Ah that's true! I totally missed that haha

Reply:
If it still gives problems, then it's okay to leave it as it is. When we fix the bug i've mentioned I'll probably fix this to take `next_key` as a value and avoid the clones.

Reply:
> Then this should work

Yeah it works, thanks a lot!

Reply:
@jedel1043 Oh it looks like the implementation of `PropertyKey` to `JsValue` conversion also does a copy :(.

https://github.com/boa-dev/boa/blob/d3a70e8a81a40a974c9ec8f9e2eec1a8554a2128/boa/src/property/mod.rs#L562-L577

Reply:
That'll get fixed when we fix the bug I mentioned, don't worry :)